### DISCUSSION TITLE
volumetric locking correction for TET10

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#25496] [Q&A Modules: Solid mechanics] [answered]
Dear Moose Community,

I have tried to activate the option "volumetric_locking_correction" (within [Modules/TensorMechanics/Master]) for a mesh that consists exclusively of TET10 elements. Here I get the following error message from Moose:
_"Volumetric locking correction should not be used for higher-order elements."_
I would have assumed that TET10 elements are suitable for this correction. Do I have a misunderstanding?

**Comment by GiudGiud** [comment]
TET10 is second order (4 vertex nodes + 6 mid-edge nodes we use for the additional dofs)
@recuero @roystgnr for the meat of the question

**Reply by roystgnr** [reply]
I'm more of a fluids guy than an elasticity guy, but I'd have assumed that Tet10 elements were suitable for the correction - https://mooseframework.inl.gov/modules/tensor_mechanics/VolumetricLocking.html shows it being applied at the weak formulation level, so whatever sins it commits ought to be somewhat element-independent.  At least in the fluids world the best stabilization schemes don't actually *hurt* anything when they're used on spaces for which they're superfluous.  I'd guess it wasn't a *necessary* correction on Tet10 ... but http://solidmechanics.org/Text/Chapter8_6/Chapter8_6.php talks about using reduced integration (the other popular way to avoid locking here) on Tet10, so maybe I guessed wrong?

**Comment by hugary1995** [comment, marked as answer]
I doubt there's volumetric locking for second order TETs (or TRI in 2D). But it's been a long time since I last looked this problem. There is a traditional Cook's membrane problem to verify whether there's locking: https://mooseframework.inl.gov/modules/tensor_mechanics/VolumetricLocking.html

The reason for the error "Volumetric locking correction should not be used for higher-order elements." must be that someone tested higher-order elements and observed that no B-bar stabilization is needed. If you are interested, please feel free to try the Cook's membrane with second order TRI6. I am interested in your findings.

**Reply by hugary1995** [reply]
Shear locking though, is a different problem and we don't currently have a good way of stabilizing it.

**Reply by jmeier** [reply]
Dear @hugary1995, @GiudGiud, and @roystgnr,

Thank you for looking into this topic and your valued input. 

@hugary1995: As far as I know Moose currently has only rectangular 4-node shell elements (#23874). If there is a TET6 in Moose in the future, the Cook's membrane test for sure is an interesting test to run.

**Reply by jmeier** [reply]
Dear all,

please find below a small test model where two "plates" made of TET10 elements get bend by a point load. One plate is coarsly meshed. The other one consists of a very fine mesh. In case of PoissonsRatio=nu=0.3 the max displacement of both plates are in good agreement. For nu=0.49 the coarsly meshed plate reacts a lot stiffer. 

A screenshot for nu=0.3 of both "plates" made from TET10 (max disp z of both plates is ca. 1.6m - color scale 0.0 - 1.6):
![grafik](https://github.com/idaholab/moose/assets/57180520/7be96afc-6cf3-4247-8bbf-b631fe196e10)

A screenshot for nu=0.49 of both "plates" made from TET10 (the coarsly meshed plate reacts stiffer - color scale 1.25-1.4):
![grafik](https://github.com/idaholab/moose/assets/57180520/0fb6eed3-be2a-48ad-bdb5-52d3ced0285d)

Disclaimer: I'm not sure if volumetric locking is causing this - or shear locking. It is currently unclear to me how to distinguish these two effects. The idea was to find out using the moose option mentioned above.

The corresponding mesh file (please remove the ".txt")
[PlateBender - OnlyVolumes.msh.txt](https://github.com/idaholab/moose/files/12656686/PlateBender.-.OnlyVolumes.msh.txt)

The input file (for nu=0.49):
```
# testing plate bending

# Model Units
modelunit_length = 'm'
modelunit_time   = 's'
modelunit_mass = 'kg' # Mg = tons; Gg = kilotons
modelunit_force  = ${raw ${modelunit_mass} * ${modelunit_length} / ${modelunit_time} ^ 2}
modelunit_pressure = ${raw ${modelunit_force} / ${modelunit_length} ^ 2}
#modelunit_acceleration = ${raw ${modelunit_length} / ${modelunit_time} ^ 2}
#modelunit_density = ${raw ${modelunit_mass} / ${modelunit_length} ^ 3}

[Mesh]
  [File]
    type = FileMeshGenerator
    file = "PlateBender-OnlyVolumes.msh"
    #show_info = true
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = false
[]

[Modules/TensorMechanics/Master]
  [./all]
    new_system = true                     # TRUE to use the new lagrangian kernel system
    formulation = total                   # controlling if the UPDATED or TOTAL Lagrangian formulation is used (default is TOTAL)
    strain = small                        # FINITE or SMALL kinematic formulations
    #volumetric_locking_correction = true

    add_variables = true

    generate_output = 'stress_xx stress_xy stress_xz stress_yy stress_yz stress_zz '
                      #'max_principal_stress mid_principal_stress min_principal_stress '
  []
[]

# ===== Boundary Conditions =====
XMinSurfaces = 'SrfVolumeL1_XMinA_1 SrfVolumeL1_XMinB_1 SrfVolumeL2_XMinA_1 SrfVolumeL2_XMinB_1'
[BCs]

  [./BoundaryZMin_DispX]
    type = DirichletBC
    boundary = ${XMinSurfaces}
    variable = disp_x
    value = 0
  []

  [./BoundaryZMin_DispY]
    type = DirichletBC
    boundary = ${XMinSurfaces}
    variable = disp_y
    value = 0
  []

  [./BoundaryZMin_DispZ]
    type = DirichletBC
    boundary = ${XMinSurfaces}
    variable = disp_z
    value = 0
  []

[]

pointLoad = ${units 100 kN -> ${modelunit_force}}

[Functions]
  [./PointLoad]
    type = ParsedFunction
    symbol_names  = 'pointLoad'
    symbol_values = '${pointLoad}'
    expression = 'pointLoad * t'
  []
[]

[DiracKernels]

  [./point_source1]
    type = FunctionDiracSource
    variable = disp_z
    function = PointLoad
    point = '+2.0 +0.0 +0.0'
    #point_not_found_behavior = ERROR
  []

  [./point_source2]
    type = FunctionDiracSource
    variable = disp_z
    function = PointLoad
    point = '+2.0 -5.0 +0.0'
    #point_not_found_behavior = ERROR
  []

[]

volPlates = 'VolVolumeL2A_1 VolVolumeL2B_1 VolVolumeL1A_1 VolVolumeL1B_1'
[Materials]

  [./volPlates_elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = ${volPlates}
    youngs_modulus = ${units 1000E3 kN/m^2 -> ${modelunit_pressure}}
    poissons_ratio = 0.49
  []

  [./volPlates_stress]
    type = ComputeLagrangianWrappedStress
    block = ${volPlates}
  []

  [./volPlates_stress_base]
    type = ComputeLinearElasticStress
    block = ${volPlates}
  []

[]

[Executioner]
  start_time = 0.0
  end_time = 1.0
  dt = 1.0

  type = Transient
  #automatic_scaling = true
  #verbose = true
  
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'
  
  nl_abs_tol = 1E-3
  #nl_rel_tol = 1E-12
  nl_max_its = 400
  
  l_tol = 1E-3
  l_max_its = 200
[]

[Outputs]
  perf_graph = true
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
I would try a coarse tet4 case with volumetric locking correction on and off and see if 
- the result with off is close to the tet10 result
- The result with the correction is better and closer to the fine mesh

**Reply by jmeier** [reply]
Is there a way to exchange the TET10 with TET4 after reading it with the [Mesh][File]? Or do I have to replace the msh file?

**Reply by GiudGiud** [reply]
You ll have to replace the file 

we are interested in the contribution of an element order converter but we don’t have it

**Reply by jmeier** [reply]
With the TET4 it gets extreme even for nu = 0.3 - but volumetric_locking_correction does not seem to have an effect. I'm doing something wrong?

(all results are u_z in meters)

For nu=0.3
| option | coarsly meshed plate | densly meshed plate |
|--------|--------|--------|
| TET4, volumetric_locking_correction=false | 0.39| 1.31 |
| TET4, volumetric_locking_correction=true | 0.39| 1.31 |


Now for nu=0.45:
| option | coarsly meshed plate | densly meshed plate |
|--------|--------|--------|
| TET4, volumetric_locking_correction=false | 0.25| 1.10 |
| TET4, volumetric_locking_correction=true | 0.25| 1.10 |

The MSH file I used:
[PlateBender-OnlyVolumes-TET4.msh.txt](https://github.com/idaholab/moose/files/12661392/PlateBender-OnlyVolumes-TET4.msh.txt)

The MSH file has actually all _nodes_ of the TET10 elements. But only a fraction of these nodes are used by the TET4-elements. Moose writes its uses TET4 and does not complain about the "floating" nodes.

**Reply by hugary1995** [reply]
volumetric_locking_correction has no effect on linear tets.
